tools/memory_estimator/memory_estimator.py

```diff
@@ -0,0 +1,200 @@
+
+    # Add FreeRTOS include dirs for Cortex-M4 to include dirs list.
+    include_dirs.append(os.path.join(freertos_src_dir, 'include'))
+    include_dirs.append(os.path.join(freertos_src_dir, 'portable', 'GCC', 'ARM_CM4F'))
+
+    # Add config files dir to include dirs list.
+    include_dirs.append(os.path.join(__THIS_FILE_PATH__, 'config_files'))
+
+    # Generate source files list.
+    if lib_name in __LIB_NAME_TO_SRC_DIRS_MAPPING__:
+        source_files=[]
+        for lib_src_location in __LIB_NAME_TO_SRC_DIRS_MAPPING__[lib_name]:
+            lib_src_full_path = os.path.join(freertos_lts, lib_src_location)
+
+            # Traverse if the location is a directory, otherwise just append it.
+            if os.path.isdir(lib_src_full_path):
+                for root, dirs, files in os.walk(lib_src_full_path):
+                    for file in files:
+                        if file.endswith('.c'):
+                            source_files.append(os.path.join(root, file))
+            else:
+                source_files.append(lib_src_full_path)
+    else:
+        raise ValueError('Library name: {} is invalid'.format(lib_name))
+
+    # Generate makefile.
+    generate_makefile_from_template(source_files,
+                                    include_dirs,
+                                    optimization,
+                                    __MAKE_FILE_TEMPLATE__,
+                                    __GENERATED_MAKE_FILE__)
+
+
+def make(compiler):
+    compiler_command='CC={}'.format(compiler)
+
+    print('---- Starting Build. Compiler: {} ---- \n'.format(compiler))
+
+    proc = subprocess.Popen(['make', '-f', __GENERATED_MAKE_FILE__, compiler_command],
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.STDOUT)
+
+    warnings = []
+    for line in iter(proc.stdout.readline, b''):
+        line = line.decode('utf-8').strip()
+        print(line)
+        if re.match('.*: warning: .*', line):
+            warnings.append(line.strip())
+
+    print('---- Build Finished ---- \n')
+    return warnings
+
+
+def calculate_size(sizetool):
+    sizetool_command='SIZETOOL={}'.format(sizetool)
+
+    proc = subprocess.Popen(['make', '-f', __GENERATED_MAKE_FILE__, 'size', sizetool_command],
+                            stdout=subprocess.PIPE)
+
+    calculated_sizes = []
+    for line in iter(proc.stdout.readline, b''):
+        line = line.decode('utf-8').strip()
+        calculated_sizes.append(line)
+
+    return calculated_sizes
+
+
+def clean():
+    proc = subprocess.Popen(['make', '-f', __GENERATED_MAKE_FILE__, 'clean'],
+                            stdout=subprocess.PIPE)
+
+    for line in iter(proc.stdout.readline, b''):
+        line = line.decode('utf-8').strip()
+        print(line)
+
+
+def remove_generated_makefile():
+    os.remove(__GENERATED_MAKE_FILE__)
+
+
+def pretty_print_warnings(warnings):
+    if len(warnings) > 0:
+        print('\n\n---- Warnings ---- ')
```

Comment:
I would recommend logging module in the next round. https://docs.python.org/3/library/logging.html. 

The benefit is you could customize log line formatting in a cleaning way, and also having more than one streams (e.g. log to both screen and file). And the file can go directly to infra for further parsing or entering into database...